test/test_wave.py

```diff
@@ -98,165 +98,104 @@ def _cos(self, val):
-        #             ("rhs_v_expected", expected_rhs[1:]),
-        #             ])
 
-    print("Approximation error:")
-    print(eoc_rec)
-    assert(eoc_rec.order_estimate() >= order - 0.5 or eoc_rec.max_error() < 1e-11)
+def get_manufactured_cubic(dim):
+    # 2D: phi(x,y,t) = cos(t-pi/4)*(x-1)^3*(x+1)^3*(y-1)^3*(y+1)^3
+    # 3D: phi(x,y,z,t) = cos(t-pi/4)*(x-1)^3*(x+1)^3*(y-1)^3*(y+1)^3
+    #                      *(z-1)^3*(z+1)^3
+    # on [-1, 1]^{#dims}
+    # (Manufactured solution)
+    def mesh_factory(n):
+        from meshmode.mesh.generation import generate_regular_rect_mesh
+        return generate_regular_rect_mesh(
+                a=(-1.,)*dim,
+                b=(1.,)*dim,
+                n=(n,)*dim)
+    sym_coords = [pmbl.var(name) for name in ["x", "y", "z"]]
+    sym_t = pmbl.var("t")
+    sym_cos = pmbl.var("cos")
+    sym_phi = sym_cos(sym_t - np.pi/4)
+    for i in range(dim):
+        sym_phi = sym_phi * (sym_coords[i]-1)**3 * (sym_coords[i]+1)**3
+    return (dim, 2., mesh_factory, sym_phi)
 
 
-@pytest.mark.parametrize("dim", [2, 3])
+@pytest.mark.parametrize(("dim", "c", "mesh_factory", "sym_phi"),
+    [
+        get_standing_wave(2),
+        get_standing_wave(3),
+        get_manufactured_cubic(2),
+        get_manufactured_cubic(3)
+    ])
 @pytest.mark.parametrize("order", [2, 3, 4])
-def test_wave_manufactured(ctx_factory, dim, order):
+def test_wave(ctx_factory, dim, order, c, mesh_factory, sym_phi):
     cl_ctx = ctx_factory()
     queue = cl.CommandQueue(cl_ctx)
 
     from pytools.convergence import EOCRecorder
     eoc_rec = EOCRecorder()
 
     for n in [4, 8, 16]:
-        from meshmode.mesh.generation import generate_regular_rect_mesh
-        mesh = generate_regular_rect_mesh(
-                a=(-1.,)*dim,
-                b=(1.,)*dim,
-                n=(n,)*dim)
+        mesh = mesh_factory(n)
 
         from grudge.eager import EagerDGDiscretization
         discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
 
         nodes = discr.nodes().with_queue(queue)
 
-        # 2D: phi(x,y,t) = cos(omega*t-pi/4)*(x-1)^3*(x+1)^3*(y-1)^3*(y+1)^3
-        # 3D: phi(x,y,z,t) = cos(omega*t-pi/4)*(x-1)^3*(x+1)^3*(y-1)^3*(y+1)^3
-        #                      *(z-1)^3*(z+1)^3
-        # Not actual solutions, but we can still apply wave operator to them
-        coord_names = ["x", "y", "z"]
-        sym_coords = [pmbl.var(name) for name in coord_names]
+        sym_coords = [pmbl.var(name) for name in ["x", "y", "z"]]
         sym_t = pmbl.var("t")
-        sym_c = pmbl.var("c")
-        sym_omega = pmbl.var("omega")
-        sym_pi = pmbl.var("pi")
-        sym_cos = pmbl.var("cos")
-        sym_phi = sym_cos(sym_omega*sym_t - sym_pi/4)
-        for i in range(dim):
-            sym_phi = sym_phi * (sym_coords[i]-1)**3 * (sym_coords[i]+1)**3
+
+        # f = phi_tt - c^2 * div(grad(phi))
+        sym_f = sym_diff(sym_t)(sym_diff(sym_t)(sym_phi)) - c**2\
```

Comment:
Is there anyplace that checks that this is small?

Reply:
So, I'm sort of looking at this from a manufactured solution point of view, where instead of solving `phi_tt = c^2 * div(grad(phi))` I'm solving `phi_tt = c^2 * div(grad(phi)) + f` where `f` comes from the solution being prescribed, and is not necessarily expected to be small. The machinery is set up here so that you can basically plug in any function you want (subject to boundary conditions), it will derive the `f` that makes the prescribed function a solution, compute numerical and symbolic RHSs for this modified equation, and then compare them. So in that sense it shouldn't really matter if `f` is small.

Reply:
OIC, that makes sense, and it's fine by me. Could you add some comments that state this explicitly? I guess you have one data point in that I didn't manage to figure this out despite some staring. (But then I also stubbornly assumed I know what was going on. :)